File: lib/engine/insurance.ts
// ============================================================================
// Aegis Vault — Insurance Fund
// Manages protocol insurance: fee routing, claims, and admin withdrawals.
// ============================================================================

import { type Fill, type InsuranceFund, BPS_DENOMINATOR } from './types';

/**
 * Create a new insurance fund with zero balance.
 */
export function createInsuranceFund(): InsuranceFund {
  return {
    balance: 0,
    totalCollected: 0,
    totalClaimed: 0,
  };
}

/**
 * Route a portion of a trade fee into the insurance fund.
 * insuranceAmount = fill.fee * insuranceCutBps / 10000
 *
 * Returns the amount routed to the fund.
 */
export function routeFeeToInsurance(
  fill: Fill,
  insuranceCutBps: number,
  fund: InsuranceFund
): number {
  const insuranceAmount =
    (fill.fee * insuranceCutBps) / BPS_DENOMINATOR;
  fund.balance += insuranceAmount;
  fund.totalCollected += insuranceAmount;
  return insuranceAmount;
}

/**
 * Claim from the insurance fund to cover bad debt (e.g., after liquidation
 * where the collateral didn't cover the loss).
 *
 * Returns the actual amount claimed (may be less than requested if fund is empty).
 */
export function claimInsurance(
  fund: InsuranceFund,
  amount: number
): number {
  if (amount <= 0) throw new Error('Claim amount must be positive');

  const actualClaim = Math.min(amount, fund.balance);
  fund.balance -= actualClaim;
  fund.totalClaimed += actualClaim;
  return actualClaim;
}

/**
 * Check if admin can withdraw from the insurance fund.
 * Admin can only withdraw amounts above a minimum buffer.
 */
export function canAdminWithdraw(
  fund: InsuranceFund,
  amount: number,
  minBuffer: number = 10000 // Default: $10,000
): boolean {
  if (amount <= 0) return false;
  return fund.balance - amount >= minBuffer;
}

/**
 * Admin withdraw from insurance fund.
 * Only succeeds if the remaining balance stays above minBuffer.
 */
export function adminWithdraw(
  fund: InsuranceFund,
  amount: number,
  minBuffer: number = 10000
): boolean {
  if (!canAdminWithdraw(fund, amount, minBuffer)) return false;
  fund.balance -= amount;
  return true;
}

File: engine/src/insurance.rs
use crate::types::{Fill, InsuranceFund, BPS_DENOMINATOR};

pub fn create_insurance_fund() -> InsuranceFund {
    InsuranceFund {
        balance: 0.0,
        total_collected: 0.0,
        total_claimed: 0.0,
    }
}

pub fn route_fee_to_insurance(fill: &Fill, insurance_cut_bps: u32, fund: &mut InsuranceFund) -> f64 {
    let insurance_amount = (fill.fee * (insurance_cut_bps as f64)) / BPS_DENOMINATOR;
    fund.balance += insurance_amount;
    fund.total_collected += insurance_amount;
    insurance_amount
}

pub fn claim_insurance(fund: &mut InsuranceFund, amount: f64) -> Result<f64, &'static str> {
    if amount <= 0.0 {
        return Err("Claim amount must be positive");
    }

    let actual_claim = if amount < fund.balance { amount } else { fund.balance };
    fund.balance -= actual_claim;
    fund.total_claimed += actual_claim;
    Ok(actual_claim)
}

pub fn can_admin_withdraw(fund: &InsuranceFund, amount: f64, min_buffer: Option<f64>) -> bool {
    let min_buffer = min_buffer.unwrap_or(10000.0);
    if amount <= 0.0 {
        return false;
    }
    fund.balance - amount >= min_buffer
}

pub fn admin_withdraw(fund: &mut InsuranceFund, amount: f64, min_buffer: Option<f64>) -> bool {
    let min_buffer = min_buffer.unwrap_or(10000.0);
    if !can_admin_withdraw(fund, amount, Some(min_buffer)) {
        return false;
    }
    fund.balance -= amount;
    true
}
